File: app/models/keyword.rb
class Keyword < ApplicationRecord

  belongs_to :project
  has_many :keywords_testcases, dependent: :destroy
  has_many :testcases, through: :keywords_testcases

  validates :project_id, :keyword, presence: true
  validates :keyword, uniqueness: {scope: :project_id}

  before_validation :convert_keyword_to_upcase

  def convert_keyword_to_upcase
    self.keyword.upcase! if keyword
  end

  def update_testcase_count
    self.update_columns  testcase_count: self.testcases.count
  end

end

File: client/angular/src/app/domain/keyword.ts
import {TestCase} from "./testcases/testcase";
export class Keyword {

  constructor(
    public id: number,
    public keyword: string,
    public projectId: number,
    public testcaseCount: number,
    public testcases: Array<TestCase>
  ) {  }

  public static create(data: any): Keyword {
    let testcases = [];
    if (data.testcases) {
      for (let testcase of data.testcases) {
        testcases.push(testcase);
      }
    }

    if (data.id == null) {
      return new Keyword(null, null, null, 0, []);
    }

    return new Keyword(data.id, data.keyword, data.project_id, data.testcase_count, []);
  }
}
